added set_namespace action

chobeat opened the pull request.

chobeat commented:
@FlorianWilhelm maybe you don't have review request notifications activated. When you have a minute can you give me your opinion on this PR?

FlorianWilhelm commented:
Sure, will do tomorrow

FlorianWilhelm commented on src/pyscaffoldext/custom_extension/custom_extension.py:
```diff
@@ -57,3 +65,13 @@ def add_entry_point(self, struct, opts):
+def set_pyscaffoldext_namespace(struct, opts):
+    PYSCAFFOLDEXT_NAMESPACE = "pyscaffoldext"
+    namespace_parameter = opts.get("namespace", None)
+    namespace_with_pyscaffoldext = ".".join([PYSCAFFOLDEXT_NAMESPACE, namespace_parameter]) \
```
What happens if someone provides an additional namespace parameter? I would actually fail here and report using the PyScaffold error logger: "Error: The namespace of an extension is `pyscaffoldext` if the user sets the namespace and if it is not `pyscaffoldext`.

chobeat replied:
the idea is that if the user provides a namespace, we put that inside the pyscaffoldext namespace. In the end we are interested that everything is inside pyscaffoldext and that the entry point is correctly specified. Additional levels in the namespace shouldn't be a problem, right?

FlorianWilhelm replied:
okay, so nested namespaces... why not. Sorry for my misunderstanding. 
@abravalheri what do you think?

chobeat commented on setup.cfg:
```diff
@@ -88,6 +88,9 @@ exclude =
     .eggs
     docs/conf.py
 
+max-line-length = 120
```
makes sense

FlorianWilhelm commented on src/pyscaffoldext/custom_extension/custom_extension.py:
```diff
@@ -57,3 +67,13 @@ def add_entry_point(self, struct, opts):
+    namespace_with_pyscaffoldext = ".".join([PYSCAFFOLDEXT_NS,
+                                             namespace_parameter]) \
+        if namespace_parameter else \
+        PYSCAFFOLDEXT_NS
```
This is quite hard to read, what about:
```python
ns_pyscaffold = PYSCAFFOLDEXT_NS
ns_param = opts.get("namespace", None)
if ns_param:  # nest provided namespace within pyscaffoldext
   ns_pyscaffold = ".".join([ns_pyscaffold, ns_param])
  
opts["namespace"] = prepare_namespace(ns_pyscaffold)
```

The trick is that you don't need extra long variable names if they are used only in a short scope.

FlorianWilhelm commented:
BTW, maybe I haven't documented that good enough in `configupdater` but instead of 
```
buffer = io.StringIO()
config.write(buffer)	      
struct[opts["project"]]["setup.cfg"] = buffer.getvalue()	         
```
you could just do
```
struct[opts["project"]]["setup.cfg"] = str(config)	    
```

chobeat commented:
> struct[opts["project"]]["setup.cfg"] = str(config)

Are you sure? `config` is a ConfigParser object. If I use `str` it will just return the default `__str__`, not the actual value. Also I'm not using `configupdater` but it sounds like something I should use.

FlorianWilhelm commented:
@chobeat, ohh I have missed that you are not using configupdater. If you use it, it would just work like I mentioned above. And I would also suggest that you do since it allows you to change and update a config file more easily and fine-grained than ConfigParser. You can import ConfigUpdater from pyscaffold.contrib, so there is no need to make it a dependency.

chobeat commented:
Alright, I will look into it

chobeat commented:
Ok, I've fougnd a fixed some bugs, reworked the tests and replaced the ConfigParser. If this looks good, I will merge and move on to the default commands

FlorianWilhelm commented on src/pyscaffoldext/custom_extension/custom_extension.py:
```diff
@@ -15,45 +19,67 @@ class CustomExtension(NoSkeleton):
+                   "{}={}.{}.{}:{}".format(opts["package"],
+                                           opts["namespace"][-1],
+                                           opts["package"],
+                                           opts["package"],
```
So the name of the extension module is also like the name of the package. I wonder if it wouldn't be better to just name it `extension.py` like it is done in our other external extensions. Personally I never name a Python module within a package like the package since a module should rather express with its name its purpose. The package name already tells you about what the name of the extension is, thus naming the module the same way is a repetition.

chobeat replied:
Yes, it was among the modifications you proposed. I planned to go over them one by one. This one is just for the namespace thing. But I can modify this here, no probs.

FlorianWilhelm commented on src/pyscaffoldext/custom_extension/custom_extension.py:
```diff
@@ -15,45 +19,67 @@ class CustomExtension(NoSkeleton):
+def add_custom_extension_structure(struct, opts):
+    custom_extension_file_content = extension(
+            get_class_name_from_opts(opts))
+    filename = "{}.py".format(opts["package"])
```
Some remark as in the comment above.

FlorianWilhelm commented on tests/test_entry_points.py:
```diff
@@ -0,0 +1,32 @@
+    assert path_exists("some_extension/setup.cfg")
+
+    config_updater = ConfigUpdater()
+    config_updater.read_file(open("some_extension/setup.cfg"))
```
It's only a unit test but still I would recommend to use a context manager here for opening the file.

FlorianWilhelm commented on tests/test_entry_points.py:
```diff
@@ -0,0 +1,32 @@
+    assert path_exists("some_extension/setup.cfg")
+
+    config_updater = ConfigUpdater()
+    config_updater.read_file(open("some_extension/setup.cfg"))
```
Same remark as above.

FlorianWilhelm commented on tests/test_entry_points.py:
```diff
@@ -0,0 +1,32 @@
+    config_updater.read_file(open("some_extension/setup.cfg"))
+    entry_point = config_updater.get("options.entry_points","pyscaffold.cli").value
+    assert entry_point == "some_extension=pyscaffoldext." \
+                          "some_extension.some_extension:SomeExtension"
```
Missing newline. Have you checked your code with flake8?

chobeat replied:
> Missing newline. Have you checked your code with flake8?

It should do it with the hooks but I've been working from different laptops so maybe I forgot to install it somewhere. Will check before merge.

chobeat commented:
ok, worked on the remarks

FlorianWilhelm commented:
@chobeat Thanks a lot! I think this can be merged.
